CBLClient/Apps/CBLTestServer-C/src/VectorSearchMethods.cpp

```diff
@@ -0,0 +1,280 @@
+            std::optional<uint32_t> subquantizers;
+            std::optional<CBLScalarQuantizerType> scalarEncoding;
+            CBLDistanceMetric dMetric;
+
+            auto* encoding = CBLVectorEncoding_CreateNone();
+            try
+            {
+                bits = static_cast<uint32_t>(body["bits"]);
+                subquantizers = static_cast<uint32_t>(body["subquantizers"]);
+            }
+            catch (...)
+            {
+                bits.reset();
+                subquantizers.reset();
+            }
+
+            try
+            {
+                scalarEncoding = static_cast<CBLScalarQuantizerType>(body["scalarEncoding"]);
+            }
+            catch (...)
+            {
+                scalarEncoding.reset();
+            }
+
+            if (scalarEncoding.has_value() && (bits.has_value() || subquantizers.has_value()))
+            {
+                throw std::invalid_argument("Cannot have scalar quantization and arguments for product quantization at the same time");
+            }
+
+            if ((bits.has_value() && !subquantizers.has_value()) || (!bits.has_value() && subquantizers.has_value()))
+            {
+                throw std::invalid_argument("Product quantization requires both bits and subquantizers set");
+            }
+
+            if (scalarEncoding.has_value())
+            {
+                encoding = static_cast<CBLVectorEncoding*>(CBLVectorEncoding_CreateScalarQuantizer(scalarEncoding.value()));
+            }
+            if (bits.has_value() && subquantizers.has_value())
+            {
+                encoding = static_cast<CBLVectorEncoding*>(CBLVectorEncoding_CreateProductQuantizer(subquantizers.value(), bits.value()));
+            }
+
+            if (metric == "euclidean")
+            {
+                dMetric = kCBLDistanceMetricEuclidean;
+            }
+            else if (metric == "cosine")
+            {
+                dMetric = kCBLDistanceMetricCosine;
+            }
+            else
+            {
+                throw std::invalid_argument("Invalid distance metric");
+            }
+
+            CBLVectorIndexConfiguration config = {};
+            config.expression = flstr(expression);
+            config.dimensions = dimensions;
+            config.encoding = encoding;
+            config.metric = dMetric;
+            config.centroids = centroids;
+            config.minTrainingSize = minTrainingSize;
+            config.maxTrainingSize = maxTrainingSize;
+            config.expressionLanguage = kCBLN1QLLanguage;
+            with<CBLDatabase *>(body,"database", [conn, collectionName, scopeName, indexName, config](CBLDatabase* db)
+            {
+                CBLError err;
+                CBLCollection* collection;
+                TRY(collection = CBLDatabase_CreateCollection(db, flstr(collectionName),  flstr(scopeName), &err), err);
+                TRY(CBLCollection_CreateVectorIndex(collection, flstr(indexName), config, &err), err);
```

Comment:
If the config.encoding is not nullptr, calling `CBLVectorEncoding_Free(config.encoding);`

Reply:
Thanks,  fixed